components/nodemanager-service/pgdb/tags.go

```diff
@@ -29,15 +50,56 @@ func (trans *DBTrans) addTags(tags []*common.Kv) ([]string, error) {
 	return tagIDs, nil
 }
 
+func (trans *DBTrans) tagExists(tag *common.Kv) (string, error) {
+	var id string
+	rows, err := trans.Query(sqlFindTagID, tag.GetKey(), tag.GetValue())
+	if err != nil {
+		return id, errors.Wrap(err, "tagExists unable to query for tag")
+	}
+	defer rows.Close() // nolint: errcheck
+	for rows.Next() {
+		err := rows.Scan(&id)
+		if err != nil {
+			logrus.Error(err)
+			continue
+		}
+	}
+	return id, rows.Err()
+}
+
+func (trans *DBTrans) nodeTagExists(nodeID string, tagID string) (bool, error) {
+	var exists bool
+	rows, err := trans.Query(sqlFindNodeTag, nodeID, tagID)
+	if err != nil {
+		return false, err
+	}
+	defer rows.Close() // nolint: errcheck
+	for rows.Next() {
+		err := rows.Scan(&exists)
+		if err != nil {
+			logrus.Error(err)
+			continue
+		}
+	}
+	return exists, rows.Err()
+}
+
 func (trans *DBTrans) tagNode(nodeID string, tagIDs []string) error {
 	links := make([]interface{}, 0, len(tagIDs))
 
 	for _, tagID := range tagIDs {
-		link := NodeTag{
-			NodeID: nodeID,
-			TagID:  tagID,
+		// check if node_tag association already exists
+		exists, err := trans.nodeTagExists(nodeID, tagID)
+		if err != nil {
+			return err
+		}
+		if !exists {
+			link := NodeTag{
+				NodeID: nodeID,
+				TagID:  tagID,
+			}
+			links = append(links, &link)
 		}
-		links = append(links, &link)
 	}
 	return trans.Insert(links...)
 }
```

Comment:
I realize this is asking for quite a bit of rework, but I feel like this database syncing stuff would be better done in the database, by making the relevant columns unique and doing `INSERT ... ON CONFLICT DO NOTHING`.

Reply:
ah good point, that would be cleaner. i'll give that a try.